### DISCUSSION TITLE
Trojanized dnSpy app drops malware cocktail on researchers, devs

### DISCUSSION THREAD

**Discussion by ndgayan** [dnSpyEx/dnSpy#76] [General]
I saw this article in the bleepingcomputer.com **https://www.bleepingcomputer.com/news/security/trojanized-dnspy-app-drops-malware-cocktail-on-researchers-devs/** about how bad actors use this amazing tool to hack into computers.

**Comment by ElektroKill** [comment]
Hi, this is sadly one of the consequences of having an open source software like this one. People with malicious intent can always just download the source code and inject malware. There is not much I can do about this as the bad actors will always find a way to counter the counter measures. I have however decided to provide hashes for dnSpy releases going forward on the releases page so it is easier to verify the download. However this won’t help much as these bad actors do a lot of get their malicious websites to the top of the search results and the information about malicious downloads won’t be seen there.

If you have any ideas on how to better circumvent this other then raising awareness and reporting the malicious websites to hosting providers and search engines then I would gladly like to hear them. 

To finish of, it’s worth mentioning that the downloads from the Releases section and the builds from the CI server of this repository are the only supported and verified dnSpyEx versions.

**Comment by mitchcapper** [comment]
This will be an issue with any popular project but I think dnSpy is a bit more at risk due to the history and current fragmentation.   

- Searching github for dnspy does not show this in the first 10 pages by default (no idea why on that one)
- Searching github for dnspyex returns only 6 results, none are you as dnspyex is only in the repo user (which it doesn't consider value in a repo search).
- Searching google for dnspy and one cannot find this repo on the first 6 pages.  Google doesn't even show the "More results from github.com ..." for me for some reason.
- The official project being discontinued without stating a successor fork means users have to try and search harder for a fork.
- Even if you know it is called dnSpyEx if you search on google for this, it autocorrects to dnspy and that has the problem of bullet 1.
- If you tell google no no you really mean dnSpyEx it lists this org first, but the first thing that looks like an actual repo is https://github.com/VNGhostMans/dnSpyEx/commits/master which is not properly forked, has its own 'latest releases' and clearly not right (see note below)
- There is only one release on this repo with a minor version increment.  Someone looking for the latest release may miss this being it.



Potential recommendations
- Un-fork this repo, and make it a source repo instead.  Github makes this quite easy now: https://stackoverflow.com/questions/38831301/how-to-un-fork-the-github-repository/66470086#66470086 .  There are several benefits with github and search engines.  Github by default excludes forks from several things.  
- Consider renaming the repo/ui to dnSpyEx.  This is already how it is referred to often, but this would make it easier to find and issues easier.  If not it may be worth just assuming spot as the "Official, Unofficial" continuation of dnSpy.   Looking at the work being done this is the massive source of development.
- Add a note to the start of the README mentioning the situation ie: "dnSpyEx is the unofficial continuation of the original dnSpy/dnspy debugger that was archived by the lead in 2020."  There is the about on the side but may be missed.
- Consider adding a note about CI beta builds with a link and disclaimer like "beta builds may have some (or all) features broken and are for testing purposes only".  There are no other official distribution methods/websites of this repo oterr than the CI system or the Releases page.

Semi-unrelated side note:
The https://github.com/VNGhostMans org is highly suspect.  At first it looks like someone who just doesn't know how to use git or github, committing up large other repos in their own commits.  They had properly forked things in the past though. In addition you look at the 'commits' being made by them (all of which are only to orgs they control) and they are odd. Why are the pinning their odd forks/mirrors and even going as far as instantly "archiving" repos to make it look like the official: https://github.com/VNGhostMans/dnSpy.    Why are they pointing in a readme to the old dnSpy as their personal copy: https://github.com/VNGhostMans/dnSpyEx/commit/5d1ab25a3286df349d6bcb079ff26ec2f04ddd91 Why are they making app releases mirroring official release names and even modifying the release descriptions: https://github.com/VNGhostMans/ScyllaHide/releases/tag/v1.4.750-7e648fb vs https://github.com/x64dbg/ScyllaHide/releases .  They also have a second org that has done very similar with many other github projects: https://github.com/Obfuscator-Collections

Maybe there is some way this is done without malicious intent. I can't think of how, even someone very 'fresh', wouldn't make these changes for innocuous reasons.  Honestly if I was trying to get malware out there stealth cloning, with some artificial history, cross linking between my repos, and faux releases that look real would very much seem like an effective way.  I didn't take the time to compare the binaries (although they at least repackaged some of them from the official archive) for anything suspect.  There is 100% the option though in the future to make a "Release" that looks legit but has one slightly modified file in it, that someone may not notice if they didn't directly compare with the actual official release.

**Reply by ElektroKill** [reply]
Hi,
First of all sorry for the rather late response. 

> * Searching github for dnspy does not show this in the first 10 pages by default (no idea why on that one)
> * Searching github for dnspyex returns only 6 results, none are you as dnspyex is only in the repo user (which it doesn't consider value in a repo search).
> * Searching google for dnspy and one cannot find this repo on the first 6 pages.  Google doesn't even show the "More results from github.com ..." for me for some reason.

This is because GitHub does not index repositories that are forked from a different one.

> * The official project being discontinued without stating a successor fork means users have to try and search harder for a fork.

I do not have any control over the dnSpy repository so I cannot really add a notice there and the original developer seems to have lost interest in dnSpy. 

> * Even if you know it is called dnSpyEx if you search on google for this, it autocorrects to dnspy and that has the problem of bullet 1.

dnSpy is searched for more often than dnSpyEx so Google may think that dnSpyEx is a typo. Although it may also be caused by the poor indexing and SEO for forks on GitHub's side.

> * If you tell google no no you really mean dnSpyEx it lists this org first, but the first thing that looks like an actual repo is [VNGhostMans/dnSpyEx@master (commits)](https://github.com/VNGhostMans/dnSpyEx/commits/master) which is not properly forked, has its own 'latest releases' and clearly not right (see note below)

I'm puzzled by the existence of this repository too. More on this later.

> * There is only one release on this repo with a minor version increment. Someone looking for the latest release may miss this being it.

There will be more releases in the future, I'm just very busy with school so I don't have much development time :)

> * Un-fork this repo, and make it a source repo instead. Github makes this quite easy now: [stackoverflow.com/questions/38831301/how-to-un-fork-the-github-repository/66470086#66470086](https://stackoverflow.com/questions/38831301/how-to-un-fork-the-github-repository/66470086#66470086) . There are several benefits with github and search engines. Github by default excludes forks from several things.

I have considered this but I haven't committed to doing it as I don't want people to be misled thinking this is some new project inspired by dnSpy but rather just a straight continuation. I know this causes a major sacrifice when it comes to SEO and discoverability of this project which may cause a lot of confusion. I hope that with time Google will get better at picking up this repository as a thing. I'm quite disappointed that the only way to improve indexing and SEO for forks is to either get more stars than the original (let's face it this will probably never happen) or unforking the repository.

> * Add a note to the start of the README mentioning the situation ie: "dnSpyEx is the unofficial continuation of the original dnSpy/dnspy debugger that was archived by the lead in 2020." There is the about on the side but may be missed.
> * Consider adding a note about CI beta builds with a link and disclaimer like "beta builds may have some (or all) features broken and are for testing purposes only". There are no other official distribution methods/websites of this repo oterr than the CI system or the Releases page.

Will do this shortly!

Regarding the VNGhostMans repository, I agree that it does look very fishy, I will try to contact the owner as to what his intent behind it is and why he has not just contributed his language changes through crowdin like everyone else. The last thing I want is for the name "dnSpyEx" to be linked with malware.

**Reply by hymccord** [reply]
I have an pro to un-forking. Submitting PRs to this repo is a PITA. When I click 'Create Pullrequest' in this repo, it automatically redirects me back to the original (now archived) repo and every time I have to edit the URL manually to submit a PR here because I haven't found a way to it without that. I think un-forking would fix that.

**Comment by mitchcapper** [comment]
Agree with your points and certainly several are out of the project's control.   

I think de-forking would help a good number of issues and any sort of concern over confusion you can address (similarly to what you just did) in the readme which should hopefully suffice for most users.   

We already know github treats forks as second class citizens and it is not hard to assume google does the same to some extent.    Aside from the lack of github features, a big one that is also disabled is searching of code from the searchbox even when on the project page.   I use git grep plenty, but there are also a good number of times I search on GH instead.   Searching the dnSpy repo right now somewhat suffices but the further this diverges the less and less useful that already inconvenient solution is.

**Comment by VNGhostMans** [comment]
Hello everyone.
After reading the discussions, it seems people are suspicious about my Github dnSpyEx project.
I can present the following:
This project is similar to the dnSpyEx project that ElectrollKill is developing 99.99% of, the other 0.01% as it says in the README.md that this project supports more Vietnamese language. Everything is the same in the dnSpyEx project. And I promise the source code is 100% clean. As for why I don't contribute to ElectroKill's dnSpyEx project, it's because I don't know much about it :D. And second, it's very difficult for newbies to know how to build it, as they don't know where the subfolders are located. This dnSpyEx project of mine (my calling is also incorrect because the original code is from ElectroKill, I have included the original link in the description) is a full project, I have updated every subfolder, as well as the build it. Also a lot of people are concerned that this project contains malicious code, but no, you see, it's completely open-source, you can check each file by yourself if you have free time. I always keep everything up to date, so this project is always updated as soon as ElectroKill updates dnSpyEx. Thanks for liking and knowing me.
As for why you can easily search my dnSpyEx repo? I don't know :D
I'm not forcing anyone to use my dnSpyEx releases, but I can assure you that my dnSpyEx repo is 100% clean. As for the reason why I didn't fork the dnSpyEx repo, it's as I stated above. Thanks!

**Reply by mitchcapper** [reply]
@VNGhostMans to be clear I was not stating anything currently contained any malicious code, also appreciative of you taking the time to respond here on some of the comments. The way you are structuring your organizations and your personal account have some things are hard to follow the logic on.   

As to why your repo performs better in searches and why you can search your repo is straightforward.   Rather than normally where one would fork a project, as you have done https://github.com/VNGhostMans/x64dbg ,  you instead have locally cloned the repositories and then copied the files into a new repo and committed them up like: https://github.com/VNGhostMans/dnSpy .  This not only destroys the true code history, but also makes you an original "source" for that repo/code.    

So if you had a minute to answer some questions I think it would help people understand your goals:

- For this project: https://github.com/VNGhostMans/dnSpy  you cloned the old repo, made the changes updated the readme put out your own release with the same version number as the official release.  You then immediately archived the repo as well.   Was there a reason to do it that way?

- For https://github.com/VNGhostMans/ScyllaHide you copied the code again, changed the line endings but nothing else, and again put a release out mirroring the official one with the same notes and no notable changes.

- For https://github.com/Obfuscator-Collections/ConfuserEx-Mod-By-Beds  not only did you do the normal stripping of the historical attribution, again your own release, but you removed the MIT license.  That is actually nearly the ONLY thing you can't do with MIT licensed code: "The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software."

- This pattern continues for many of the other repositories under you and your organization.   If this is for your private use why not make the repositories private?  If it is for public consumption why are you creating seemingly official releases that are clones of the actual releases?

 - I think one of the points you were making above was that you were making your own release of dnSpyEx as it can be hard to build and so you were doing it for those who could not.   This does not quite explain the releases of the other repo's where there were no changes.  

 - On your public dnSpyEx repo you have force pushed over the previous copy you had 5 hours ago, erasing your own history.  This is apparent in the link I posted above is now broken.

 This is just a select review of a few of the repositories you and your org have, I did not check them all nor actually compare the release binaries, but if you can answer these questions I think it would be easier for us to understand your intentions. 

Finally, if you are going to put out a public product different from the official, you likely should rename them and not be directly copying most of the details of the official repo.  Even if without malicious intent it can unquestionably lead to confusion.

**Reply by VNGhostMans** [reply]
This is my own repository, not that of any group.
Regarding Obfucator-Collections, I personally just want to collect the publicly available .NET encryption software, and organize them in one place. Thus users will be able to easily search without spending much time. As for why most of my repos don't fork from the author because some repos come with subfolders, which makes it difficult for new users to find. So I decided to download and update the subfolders, and include the author's release. Why do I make it public instead of private? Because it's not a product that I created :D. And as you say I don't leave the source and change the README.md? I can tell you the following, if I decide to edit the README.md content, I will add the original link in the "About" section. If the repo is a clone, I won't leave the original link. Speaking of which, you might be wondering why I don't leave the original link for the cloned repos? Because I did not change any of the contents of that repo. So users who can search keywords will simultaneously see my cloned repo and its original repo.
I hope that my public project repos will be trusted by you instead of doubting it.
Previously, there was a malicious version of dnSpy, they even created a website to trick users into downloading. And I confirm this is also the reason why you are so suspicious of my dnSpyEx project.
Why is my dnSpyEx project on Github often removed and re-uploaded right after? Since I'm a clumsy person, I took it down and re-uploaded it while copying ElectroKill's updated changes.
Again, I want to make sure I'm not a bad person, my repos are just for the sake of making it easy for newbies to build. I hope you guys trust and use the releases I put out instead of suspecting it as malware. I'm also not happy when I'm discussed by you guys.

**Reply by mitchcapper** [reply]
Again thank you for taking the time to respond with such detail. Sorry the discussion has made you unhappy, as you pointed out there have been malicious copies of apps like dnSpy put out before.  I also want to say that any assistance towards open source projects, organization, documentation, or translation is fantastic and very welcomed by most.  I don't want to dissuade you from trying to help in those areas.  Any implication of less than good intentions is based solely on some of the suspect actions taken and based on your comments here likely untrue.  

When I referred to the organization I simply meant your Obfucator-Collections org, but understand it is all just your repositories.  Based on some of your comments and style of cloning, I think you may be somewhat newer to github and some of the best standards and practices.   It sounds like you have a personal use of these collections along with a public goal of allowing other newer users to find things more easily.  These are great goals, but should likely be done in different ways.

A) For your personal use do whatever you like however you like. Here are some things that may be useful:  

- For organizing repos you want to remember GitHub has stars  clicking the star on a repo will add it to your personal Star list.  If you have many different things you star you can create "Lists" now under stars ( https://github.com/VNGhostMans?tab=stars ) and have a list of ".NET encryption apps" and another of ".NET obfuscation tools" for example.

- I highly suggest using GitHub forking of repositories you want to make changes in.  You can do this very easily from within github by clicking the "Fork" button on the upper right.  Not only does this preserve history, but it makes it far easier to update. GitHub now even has a button you can click on your forks to synchronize over new commits from the source repo.  This also makes it far easier for you to remember what you changed, otherwise if you look at a repo it is much harder to know what you did vs what was original code.

- If you want to continue to copy repos in your current style and make your own releases I highly recommend making the repositories private otherwise they will likely be more of a dis-service to others than a help.  There is no reason to make a repo public just because it is originally a public repo.  

B) For publicly helping users there are some fantastic ways to do so.  Your github stars and lists can be publicly linked to (private lists also coming).  Rather than manually copying and the repos into your collections put together collection/summary style repos with documentation.  For example: https://github.com/nforest/awesome-decompilation  or https://github.com/rogerwcpt/xamarin-mvvvm-frameworks.  I am not sure what your comment that most repos `come with subfolders, which makes it difficult for new users to find` means,  but with a summary repo like one of those you could deep link into whatever sub folder you want for the project.  

The reasons for not doing what you are currently doing are numerous:

- You have constant maintenance of your many repositories to make sure they stay in sync with the actual repository.  If you don't users who use your version may be using an out of date version without knowing it.

- By destroying repository history your users would be unable to look back at how a file came to its current state or who wrote a piece of code.  

- That same destruction makes it much harder for users who modify the code to contribute back to the original project.  Normally if a user clones a repo and makes changes they can easily create patches or pull requests.  If they clone your repo best case if they could create a patch and it applies cleanly to the official repo (assuming they know enough to do that) and then they can continue like normal.   If the repositories are not perfectly in sync (or they did the work and commits made since then) without proper git history merging the changes in becomes more laborious.   In addition, several of your clones adjust whitespacing on files which would make most patch processes fail.

- Again, simple user confusion about official repository source.  For example, you allow PR and bug reports to be filed in your clones, however these should be done in the original repo (especially when you have not made any changes to said repo). Even if you add a link back to the original repositories it is easy for users to miss these things.

-  There are benefits to why github specifically treats forks as they do.  It helps to reduce noise in their search engine, it makes it easier to evaluate what changes have been made by what people across many forks, and it added a standardized way for users to quickly identify if a repo is a fork or not.  When you post repos as you do, all these things are removed.

- Your remove and re-create every update for repositories will cause any user who does clone your repo to break when they try to update.


Many repos now do have CI with artifacts for people who don't want to build themselves.  If you think it is of great service to build binaries for the public from 3rd party repos consider putting binary only repos together.  Make sure to be very clear in labeling then un-official builds and mention the exact git commit you are building off of.  That will make it far easier for users to understand what is going on.

There are no hard and fast rules, these are my personal suggestions.  I think despite your intentions you are likely to do a disservice to less experienced users

Finally, as mentioned above, you need to be carefuly not remove copyright and license terms.  If you do so you could find yourself violating copyright law.

**Reply by VNGhostMans** [reply]
Thank you for understanding me and giving useful advice.
I am a Github newbie, and my purpose of using Github is simply to store documents, files that I create, software that I collect on Github.
Honestly, you are concerned that the project repo on my Github is not fork but a copy, and you are afraid that they are edited and may contain viruses, trojans... But I don't do that, do you think if I had bad intentions, would I post the source code? Or at least I will invent some features to trick users into downloading files with viruses. I know how to fork a repo, but I don't like it very much because: After fork a repo, the subfolders included in the folder are not synchronized making them discrete, nor are the releases included in the original repo Attached. As I mentioned above I am a perfectionist guy, so I decided to copy without fork it.
And also thanks for your advice, I also have my mistake for not leaving information about the author's source. I will add the source link in the "About" section when I have some free time. Thank you, thank you to everyone in this discussion.

**Reply by mitchcapper** [reply]
Again, while I think such a setup could be useful to a malicious user, I am not implying that was your goal here. I believe you are trying to help users, and was trying to explain what is likely the best way to do so.   My concern here is that you are likely to cause more confusion and cause disservice to users who use your collections rather than help them as you desire.  Still if your primary goal is to store documents for your personal use just make the repos private and you avoid any confusion to outside parties.

It is important to note there is 0 requirement for the source code to match the binary release files, so publishing one set of sources and another set of binaries is easily done.   

To address the issue of forking not copying releases nor automatically synchronizing this is true.  I assume you are talking the repo as a whole, if you mean by sub-folders sub-modules it does infact keep them in a clone, ie: https://github.com/mitchcapper/dnSpy/tree/master/Libraries .   If you manually clone the repo you also do not have it automatically staying in sync, and you 100% are not having releases cloned.   Forking makes it dead simple to sync, it is a button press within github or two commands from a command line.   For you to currently sync I am not even sure the process maybe:
1) Create a fresh clone of the official repo, apply your changes, copy that to a new git instance, and commit up to your changes.
2) (Faster but not sure you would go the route) Pull changes from the official upstream repo into your local correct git mirror of them.  Copy all those files to a new repo and put that repo on GH.
Neither would seem faster than the single button on GH.   

As for releases even with forks you can create releases so you could continue to do releases the way you have been (but I would highly recommend tagging them as unofficial as well to make it clear, especially if you are not copying the official binaries).

**As to why it doesn't copy releases? My take would be:** 
It doesn't copy releases as a clone (either a direct fork or your style) of a repo should _not_ have copies of the official (or self made to theoretically mirror the official) binaries.  Users should download those from the original source.   If you are going to release a modified binary you likely should rename the project or make it clear the binaries put out are unofficial binaries with X added.

Finally, in terms of the repo licensing example.  There are no conditions under which you are allowed to completely remote an MIT license, simply linking to the original author would not suffice to meet the conditions.  The MIT license is one of the most permissive licenses out there so this would be considered a very low bar.  If you plan to alter/remove licensing information you will want to read the terms carefully as many are much more copy-left in their requirements. 

I understand you are a GH newbie, I do not believe you have malicious intent, I am trying to suggest ways you might be able to more efficiently help yourself and other users without accidentally causing a detriment to those who stumble upon your repos.

**Reply by VNGhostMans** [reply]
Ahhh, so now what I need to do is label the current releases available on my Github as unofficial releases. I think this will let the user know that the release available on my Github is a clone from the original project. To the question that people suspect that I am the bad guy, the versions I release contain malicious code, I would like to present the following:
- My Github is geared towards new Github users like me.
- My Github is a repository, not a place where I work.
- All my Github projects are clones and of course attached releases are clones only.
- IMPORTANT IS MY GITHUB PROJECTS ARE COMPLETELY CLEAN, IF YOU ARE IN DOUBLE, YOU CAN ANALYZE IT IT YOURSELF, I WILL NOT HAVE TO EXPLAIN LONGLY BECAUSE YOU WILL THINK I'M ONLY REASONING IT MY LICENSE ACTION.
Any questions you and others can ask me, I will answer in my spare time.
Postscript:
- I'm really annoyed when my Github projects are discussed and suspected of containing malicious code. I know you're just doing this out of precaution and I don't blame you at all. What I need to do is explain. And I already explained.
- And I hope people can understand me, believe in it instead of doubting and gossiping about my projects. Thanks. Have a nice day.
- My English is not good, I use Google translate to be able to communicate with you, so if there are mistakes, please forgive me.
VNGhostMans

**Reply by mitchcapper** [reply]
Thanks again for the reply, and I think I have provided more than a welcome amount of feedback.  As you mentioned it is your github account and projects, so run them as you like. Wish you the best and thanks for helping contribute!

**Reply by VNGhostMans** [reply]
More and more bad guys take advantage of the fact that large software has open source code on Github, from which those bad guys can insert malicious code and spread it to users. It's not wrong for you to suspect me of being a bad guy, because you don't know who I am, and what is my purpose in not fork the repo like a normal person would. Thank you for understanding what I'm doing. Thanks. Have a nice day.

**Reply by Lynnolium** [reply]
MD5 and SHA1 signatures help verify that the build/binaries downloaded are legit. Unfortunatly very few outside of the serious coder bother to do that check.

**Reply by mitchcapper** [reply]
Yeah but if the repo you are looking at you think is the official repo, checking the hashes they list won't help you.
